### ISSUE TITLE
How to access the transaction ID commit?

### ISSUE THREAD

**Issue by jabalsad** [awslabs/amazon-qldb-driver-go#57]
Hi,

What's the best way to obtain the QLDB transaction ID after committing?

**Comment by marcbowes** [comment]
The overall model in QLDB is that the call to `execute` can return whatever you want. The function literal you provide is given a `Transaction`:

https://github.com/awslabs/amazon-qldb-driver-go/blob/bfe30eb197e5ec640665935f22ebf077b03beb75/qldbdriver/qldbdriver.go#L96
https://github.com/awslabs/amazon-qldb-driver-go/blob/bfe30eb197e5ec640665935f22ebf077b03beb75/qldbdriver/transaction.go#L26

This interface does not currently support returning an id, even though the underlying struct has it:

https://github.com/awslabs/amazon-qldb-driver-go/blob/bfe30eb197e5ec640665935f22ebf077b03beb75/qldbdriver/transaction.go#L37

We'll get back to you with a suggestion on how best to do this. In the meanwhile, it'd be helpful to know the use-case as there may be something else to consider.

**Comment by jabalsad** [comment]
Hi @marcbowes  :-) Thanks for getting back to me. Indeed, I saw that the underlying struct has the private field `txnId`. My understanding is that the transaction id should be obtained once a transaction has been committed? (i.e., the function literal has executed with success?).

Regarding our use-case, we also have the concept of a "transaction" in our data model, and it maps onto a QLDB transaction. Our transactions consist of a collection of documents. We'd like to be able to leverage this identifier in the same way that we leverage the QLDB document ID. However, it's also possible for us to simply generate our own identifier to group documents within a transaction and index on this field instead. It's not clear to me that there's a benefit one way or the other, other than reducing the number of indexes we create.

**Comment by marcbowes** [comment]
> My understanding is that the transaction id should be obtained once a transaction has been committed? (i.e., the function literal has executed with success?).

Definitely - because of retries. The functions need to be "pure and idempotent", i.e. can be retried without side-effects. On each attempt, you'll have a new transaction id, so it's important to use the one that actually won out.

> Regarding our use-case, [..] other than reducing the number of indexes we create.

I'm not clear on this. Are you planning on inserting the transaction id into the document and then indexing it? And you'd do that in a second transaction? If so, you can query the transaction id by querying the metadata:

```
us-west-2:shell-dev> create table example
{
  tableId: "LtMNccMTOzH5nGz8LfOu1A"
}
1 document in bag (read-ios: 1, server-time: 36ms, total-time: 166ms)
us-west-2:shell-dev> insert into example {'foo':'bar'}
{
  documentId: "2f0sjqtpyIz9cx9PbAKk3Z"
}
1 document in bag (read-ios: 0, server-time: 19ms, total-time: 133ms)
us-west-2:shell-dev> select metadata.txId from _ql_committed_example where metadata.id = '2f0sjqtpyIz9cx9PbAKk3Z'
{
  txId: "3WF7xMuYtr9D6Qsy6pwNkT"
}
1 document in bag (read-ios: 1, server-time: 29ms, total-time: 131ms)
```

**Comment by marcbowes** [comment]
> My understanding is that the transaction id should be obtained once a transaction has been committed? (i.e., the function literal has executed with success?).

The id is actually available after the transaction is started (it is a "speculative transaction" at that point), just not available in the interface.

**Comment by jabalsad** [comment]
> Definitely - because of retries. The functions need to be "pure and idempotent", i.e. can be retried without side-effects. On each attempt, you'll have a new transaction id, so it's important to use the one that actually won out.
That makes sense. 

> I'm not clear on this. Are you planning on inserting the transaction id into the document and then indexing it? And you'd do that in a second transaction? If so, you can query the transaction id by querying the metadata:
We wouldn't persist the transaction ID separately, just return it to our API callers in the response. However, the more I think about this, the more it makes sense for us to generate our own ID to group these records by and just burn an index on it.

**Comment by jabalsad** [comment]
Feel free to close this ticket for now if you don't feel like it's necessary to expose this field in the interface just yet.

**Comment by alpian** [comment]
Hi @jabalsad, I see no reason not to expose it and your use case seems entirely reasonable. I'm working with the team now to make this change and will come back to this issue once we have figured out how and when we can do that. You are, of course, welcome to raise a PR for the change you need too, if you prefer :).

**Comment by alpian** [comment]
Looks like @justing-bq has some time to do this, so we will make this change as soon as we can :).

**Comment by justing-bq** [comment]
This issue has been addressed and will be included in the next release.

**Comment by jabalsad** [comment]
Thanks for the fast turnaround!

**Comment by justing-bq** [comment]
Hi @jabalsad,

We just released [driver v2.0.2](https://github.com/awslabs/amazon-qldb-driver-go/releases/tag/v2.0.2) which addresses this issue.
